src/llvmir2hll/graphs/cfg/cfg_traversal.cpp

```diff
@@ -163,12 +176,78 @@ bool CFGTraversal::getCurrRetVal() const {
 * @param[in] node Node to be traversed.
 * @param[in] stmtIter Iterator to a statement in @a node to be checked.
 *
-* If @a stmtIter equals @c node->stmt_end(), the function traverses all
-* successors of @a node.
+* This function traverses the entire CFG rooted at @a node, starting with the statement
+* in @a stmtIter.
 */
 bool CFGTraversal::performTraversalImpl(ShPtr<CFG::Node> node,
 		CFG::stmt_iterator stmtIter) {
-	while (stmtIter != node->stmt_end()) {
+	// The stack nodes consist of the current CFG node, a statement iterator storing the current
+	// place in the node's statements, and an iterator storing the current place in the node's
+	// successor list.
+	// Statement place is only needed because we may be called with a statement iterator
+	// that doesn't point to the beginning.
+	using NodeType = std::tuple<ShPtr<CFG::Node>, CFG::stmt_iterator, CFG::succ_iterator>;
+	std::stack<NodeType> visitStack;
+	visitStack.push({node, stmtIter, node->succ_begin()});
+
+	// This logic is a bit complicated because it's normally part of a depth first iterator.
+	auto retVal = getEndRetVal();
+	do {
+		// Note that we directly mutate the successor iterators that are on the
+		// stack, because we can't pop it until we are done.
+		auto node = std::get<0>(visitStack.top());
+		auto stmtPlace = std::get<1>(visitStack.top());
+		auto &succPlace = std::get<2>(visitStack.top());
+
+
+		// First visitSingleNode will visit all remaining statements in the node.
+		retVal = combineRetVals(retVal, visitSingleNode(stmtPlace,
+								node->stmt_end()));
+		if (stopTraversal) {
+			break;
+		}
+		// Now push the next successor on the stack to make it get
+		// visited.  We have to track whether we push an element because
+		// they may have all been visited already.
+		bool pushedElement = false;
+		while (succPlace != node->succ_end()) {
+			auto succNode((*succPlace)->getDst());
+			// This mutates the actual in-stack succ iterator
+			++succPlace;
+			// Push this node on the stack to be visited if it
+			// hasn't been visited and isn't empty.
+			if (succNode->stmt_begin() != succNode->stmt_end() &&
+			    !hasItem(checkedStmts, *(succNode->stmt_begin()))) {
+				pushedElement = true;
+				visitStack.push({succNode, succNode->stmt_begin(),
```

Comment:
Similarly as in one of the above comments, consider calling `emplace()` instead of `push()` here.